allocation: time compute rounds by start time over last convergence

schase-es opened the pull request:
In the DesiredBalanceAllocator, there are periodic log messages that warn of long-running allocation balancing rounds. These track the number of loop iterations and the amount of time passed since the last convergence, and log when either exceeds a limit. A previous effort (4c979aa, #100850) persisted these metrics across compute runs so that a sequence of cluster state changes did not disrupt these warnings. However, the modified time-based threshold compares the current time against the last convergence, instead of the round's start time. If enough time has passed between balancing rounds, this produces a warn-level message that misrepresents the round's compute time. This change continues to include the time since last convergence, but uses the time since compute began as the log threshold and as the time since resumption.

Fixes: ES-13327

schase-es committed:
allocation: time compute rounds by start time over last convergence

In the DesiredBalanceAllocator, there are periodic log messages that warn of
long-running allocation balancing rounds. These track the number of loop
iterations and the amount of time passed since the last convergence, and log
when either exceeds a limit. A previous effort (4c979aa, #100850) persisted
these metrics across compute runs so that a sequence cluster state changes did
not disrupt these warnings. However, the modified time-based threshold compares
the current time against the last convergence, instead of the round's start
time. If enough time has passed between balancing rounds, this produces a
warn-level message that misrepresents the round's compute time. This change
continues to include the time since last convergence, but uses the time since
compute began as the log threshold and as the time since resumption.

Fixes: ES-13327

Commit:
[CI] Auto commit changes from spotless

schase-es commented:
This is for:
* https://elasticco.atlassian.net/browse/ES-13327
and
* https://elasticco.atlassian.net/browse/ES-12824

The main change is that when the time-based non-convergence log decision is done, it compares the time since compute started, instead of the time since last convergence.

Throughout both convergence and non-convergence logging, all phrases like "still not converged after" or "converged after" uses the time since compute started, instead of the time since last convergence. 

Similarly, all messages end with "since the last convergence <time-since-convergence> ago" have switched over to this, instead of saying "since the last resumption" which used the time since compute started.

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputerTests.java:
```diff
@@ -1428,7 +1430,7 @@ private void assertLoggerExpectationsFor(Runnable action, MockLog.LoggingExpecta
     public void testLoggingOfComputeCallsAndIterationsSinceConvergence() {
         final var clusterSettings = new ClusterSettings(
-            Settings.builder().put(DesiredBalanceComputer.PROGRESS_LOG_INTERVAL_SETTING.getKey(), TimeValue.timeValueMillis(5L)).build(),
+            Settings.builder().put(DesiredBalanceComputer.PROGRESS_LOG_INTERVAL_SETTING.getKey(), TimeValue.timeValueMillis(1L)).build(),
```
I don't think we should do this - at `5ms` some time can pass without seeing another log message (and that's what we want) but at `1ms` we expect a log message every tick of the clock.

DaveCTurner reviewed (commented):
Thanks Simon, I left a couple of comments mostly around testing.

schase-es commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputer.java:
```diff
@@ -326,7 +326,7 @@ public DesiredBalance compute(
         final long timeWarningInterval = progressLogInterval.millis();
         final long computationStartedTime = timeProvider.relativeTimeInMillis();
-        long nextReportTime = Math.max(lastNotConvergedLogMessageTimeMillis, lastConvergedTimeMillis) + timeWarningInterval;
+        long nextReportTime = computationStartedTime + timeWarningInterval;
```
This is an intentional change from me, and I adjusted the testing to match. (This is the setting change that shortens the log interval to 1 millisecond below.)

The issue I'm pointing out is subtle, and this code is hard to read. These messages have caused a few issues, so I want to get this right.

The period log interval setting was initially built to log periodically when the compute loop was taking too long. In 4c979aa, this facility was extended so that its initial report is based off the time since the last convergence. So whenever compute starts again some time after a convergence and more than the period interval has passed, we see a log message that it still hasn't converged.

The ticket is about fixing the log message so the time duration reported in this message reflects the amount of time spent in the compute loop. I could leave this in place, but there will still be two issues:

1. when we restart compute after a long time away (which may be normal), we'll still see a log about allocation still not being converged at the end of the first compute loop, even though nothing is wrong. The duration will be accurately reported, which is a positive step.

2. an early exit caused by cluster state changing will always be logged at debug.

To fix both of these, we should:

- start tracking the time since the first call into compute after a convergence, and use this as the base for the initial log period. This is more significant than the last convergence, because it's when allocation starts trying. This whole idea may be based on a misunderstanding of how and when allocation works: if it's constantly running, then there won't be any difference between the end of the last convergence and the start of the next balance.

- whenever the cluster state changes and more than the log period has passed since allocation last started up, the early exit message should log at info

Let me draft something so you can see this in practice.

DaveCTurner replied:
I see ok, so are you saying that you're proposing to (effectively) revert #126008 in this PR and then address #137020 in a separate PR? I don't think we'd want to do that as two separate steps because we might cut a release in the intermediate, regressed, state.

schase-es committed:
Now introducing firstComputeSinceConvertedTimeMillis... in a can!

schase-es committed:
Adding Yang's test

schase-es committed:
Test changes

schase-es committed:
Edited log messages to group duration and iterations/rounds logically, into
information about this past round, the time since compute began, and the time
since convergence.

schase-es committed:
Merge branch 'main' into ES-12825_long-balance-computation-log-message

schase-es committed:
Updated timing fix and reverted log message changes

firstCompute time is now calculated whenever the last run converged.

schase-es committed:
Merge branch 'main' into ES-12825_long-balance-computation-log-message

schase-es committed:
Merge branch 'main' into ES-12825_long-balance-computation-log-message

schase-es committed:
Merge branch 'main' into ES-12825_long-balance-computation-log-message

DaveCTurner reviewed (commented):
I left two tiny tiny comments but otherwise this looks exactly what we need.

schase-es committed:
Rename of first compute time

schase-es committed:
Move of test to group by log check

schase-es committed:
Fixing up broken renames

schase-es committed:
Merge branch 'main' into ES-12825_long-balance-computation-log-message

schase-es committed:
Merge branch 'main' into ES-12825_long-balance-computation-log-message

schase-es committed:
Merge branch 'main' into ES-12825_long-balance-computation-log-message

schase-es committed:
Merge branch 'main' into ES-12825_long-balance-computation-log-message

schase-es committed:
Merge branch 'main' into ES-12825_long-balance-computation-log-message

DaveCTurner reviewed (approved):
LGTM
